### PR TITLE
Add update_concurrency option to control the parallelism of shard updates

### PR DESCRIPTION
**Pull request by dzhao**
### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by timvisee — COMMENTED** [review]
Nice work! I have some comments, please see below.

Also, I'm wondering if you've tested this implementation on a large cluster?

Happy to approve once we resolve the open comments.

File: lib/collection/src/shards/replica_set.rs [hunk 1]
```diff
@@ -1580,19 +1580,19 @@ impl ShardReplicaSet {
                 )));
             }
 
-            let mut remote_futures = Vec::new();
+            let mut update_futures = Vec::new();
```
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/shards/replica_set.rs]: Resolved in a commit on your branch: 4e78be0218d826df61c57e0f992fd5d8f27230c6

File: lib/collection/src/operations/shared_storage_config.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@ pub struct SharedStorageConfig {
     pub handle_collection_load_errors: bool,
     pub recovery_mode: Option<String>,
     pub search_timeout: Duration,
+    pub update_concurrency: Option<usize>,
```
[the remark this answers is not in the document]
Comment (dzhao) [reply lib/collection/src/operations/shared_storage_config.rs]: makes sense. changed

**Comment by dzhao** [comment]
yes I have tested this with my 33 node cluster. I haven't seen any negative impact. I don't see read performance improvement since you need another PR to bypass the local shard read.

File: config/config.yaml [hunk 3]
```diff
@@ -108,6 +108,9 @@ storage:
     on_disk: false
     # Custom M param for hnsw graph built for payload index. If not set, default M will be used.
     payload_m: null
+  # Maximum number of concurrent updates to remote replica sets on other peers.
```
Comment (agourlay) [inline config/config.yaml:111]:
```suggestion
  # Maximum number of concurrent updates to replica sets.
```

It seems that a local replica set is also potentially targeted.

File: lib/storage/tests/integration/alias_tests.rs [hunk 4]
```diff
@@ -55,6 +55,8 @@ fn test_alias_operation() {
         handle_collection_load_errors: false,
         recovery_mode: None,
         async_scorer: false,
+        update_concurrency: Some(NonZeroUsize::new(2).unwrap()),
```
Comment (agourlay) [inline lib/storage/tests/integration/alias_tests.rs:58]: I don't think this should get in as there is no assertion associated with it.

File: config/config.yaml (hunk 3 above)
Comment (timvisee) [reply config/config.yaml:111]: You're right. Sorry for not catching this in my initial suggestion to add this comment.

File: lib/storage/tests/integration/alias_tests.rs (hunk 4 above)
Comment (timvisee) [reply lib/storage/tests/integration/alias_tests.rs:58]: Though I have no clear idea on how to test this then, I do think it's good that we at least cover it somewhere.

Any ideas on a test?

**Review by generall — APPROVED** [review]

**Review by timvisee — APPROVED** [review]

**Pull request merged**